### PR TITLE
Add support for save and load mkldnn modules

### PR DESCRIPTION
**Pull request by bddppq**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #20820 Add mkldnn sigmoid operator&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D15455866/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #20800 Enable torch.jit.trace for mkldnn modules&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D15447892/)
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#20799 Add support for save and load mkldnn modules**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D15447891/)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/20799

Differential Revision: [D15447891](https://our.internmc.facebook.com/intern/diff/D15447891/)

### CODE REVIEW

File: test/test_mkldnn.py [hunk 1]
```diff
@@ -231,12 +237,27 @@ def test_linear(self):
...
             mkldnn_linear = mkldnn_utils.to_mkldnn(copy.deepcopy(linear))
             self.assertEqual(
                 linear(x),
                 mkldnn_linear(x.to_mkldnn()).to_dense())
 
+            self._test_serialization(mkldnn_linear, (x.to_mkldnn(),))
+
+    def _test_serialization(self, module, inputs):
+        if IS_WINDOWS or IS_SANDCASTLE:
+            # TemporaryFileName is not supported in Windows and Sandcastle
```
Comment (dzhulgakov) [inline test/test_mkldnn.py]: there's a helper function for it already in test_jit.py. Maybe pull it out to common_utils.py and use here?

File: test/test_mkldnn.py [hunk 2]
```diff
@@ -231,12 +237,27 @@ def test_linear(self):
...
+            return
+
+        with tempfile.NamedTemporaryFile() as f:
+            torch.jit.save(module, f.name)
+            f.flush()
+            loaded = torch.jit.load(f.name)
+            self.assertEqual(
```
Comment (dzhulgakov) [inline test/test_mkldnn.py]: do you need to explicitly verify here that params are in mkldnn layout? or just passing input arg in mkldnn layout is enough?

File: torch/utils/mkldnn.py [hunk 3]
```diff
@@ -1,43 +1,149 @@
...
-        if t.is_floating_point():
-            return t.to_mkldnn()
+class MkldnnLinear(torch.jit.ScriptModule):
+    def __init__(self, dense_module):
+        super(MkldnnLinear, self).__init__()
+        self.register_buffer('weight', dense_module.weight.to_mkldnn())
+        if dense_module.bias is not None:
+            self.register_buffer('bias', dense_module.bias.to_mkldnn())
+        else:
+            # TODO: Remove this once ScriptModule supports registering None buffer
```
Comment (dzhulgakov) [inline torch/utils/mkldnn.py]: is there an issue for it?

File: torch/utils/mkldnn.py [hunk 4]
```diff
@@ -1,43 +1,149 @@
...
+            self.register_buffer(
+                'bias',
+                torch.zeros([dense_module.weight.size(0)], dtype=torch.float).to_mkldnn())
+
+    @torch.jit.script_method
+    def __getstate__(self):
+        return (self.weight.to_dense(), self.bias.to_dense())
```
Comment (dzhulgakov) [inline torch/utils/mkldnn.py]: so there's no way to return None here so that serialized format doesn't change in the future when you do the TODO. You probably could have a bool attribute 'has_bias'.

File: torch/utils/mkldnn.py [hunk 5]
```diff
@@ -1,43 +1,149 @@
...
+            self.groups)
+
 
+class MkldnnBatchNorm2d(torch.jit.ScriptModule):
+    __constants__ = ['exponential_average_factor', 'eps']
+
+    def __init__(self, dense_module):
+        super(MkldnnBatchNorm2d, self).__init__()
+
+        assert(not dense_module.training)
```
Comment (dzhulgakov) [inline torch/utils/mkldnn.py]: put better error messages? and I'm not sure what the convention is - I thought it's less of asserts and more of explicit `raise ValueError`

File: torch/utils/mkldnn.py [hunk 6]
```diff
@@ -1,43 +1,149 @@
...
+        assert(dense_module.track_running_stats)
+        assert(dense_module.affine)
+
+        if dense_module.momentum is None:
```
Comment (dzhulgakov) [inline torch/utils/mkldnn.py]: would this also be a TODO?

**Review by dzhulgakov — COMMENTED** [review]
Btw, what happens if I just try to .save() a mkldnn tensor? does it magically convert to_dense? or do we produce a nice error message?

File: torch/utils/mkldnn.py (hunk 3 above)
Comment (bddppq) [reply torch/utils/mkldnn.py]: ScriptModule has overridden `register_buffer` to the C `_register_buffer` (same for *_parameter) which doesn't accept None as the argument. IIUC other script modules [store None as attribute](https://github.com/pytorch/pytorch/blob/796e359/torch/jit/__init__.py#L1534-L1536) to work around it. However, since this module has `__getstate__` and `__setstate__`, `export` will [skip serialization of any attributes](https://github.com/pytorch/pytorch/blob/796e3596014b0f684645aa58cb0fbd806a65b070/torch/csrc/jit/export.cpp#L853-L864).

Comment (dzhulgakov) [reply torch/utils/mkldnn.py]: it's fine that we're skipping serialization of attributes. So what prevents us making it an attribute and handling it explicitly in `__get_state__/__set_state__`? or is there a specific reason it has to be a buffer?

and my original question was whether there's a GH issue to track supporting nullable buffers

File: torch/utils/mkldnn.py (hunk 4 above)
Comment (bddppq) [reply torch/utils/mkldnn.py]: Makes senses, will change it

File: torch/utils/mkldnn.py (hunk 6 above)
Comment (bddppq) [reply torch/utils/mkldnn.py]: sorry could you clarify why? (this is the same logic as in torch.nn.BatchNorm2d)
Note unlike the bias, since we don't need to serialize `momentum` so it doesn't have the same problem.

File: torch/utils/mkldnn.py (hunk 5 above)
Comment (bddppq) [reply torch/utils/mkldnn.py]: Yeah I think `ValueError` makes more senses here

File: test/test_mkldnn.py (hunk 2 above)
Comment (bddppq) [reply test/test_mkldnn.py]: It's sufficient to verify it by passing mkldnn tensor as input. For modules that don't have parameters, it's not worth testing serialization anyway. Also right now there is no way to tell if a tensor is mkldnn layout at python land.

File: torch/utils/mkldnn.py (hunk 3 above)
Comment (bddppq) [reply torch/utils/mkldnn.py]: > my original question was whether there's a GH issue to track supporting nullable buffers

There is one, but in this lockdown repo.

Comment (bddppq) [reply torch/utils/mkldnn.py]: >  So what prevents us making it an attribute and handling it explicitly in __get_state__/__set_state__? or is there a specific reason it has to be a buffer?

So I don't think it has to be a buffer, the problem I had run into is the **parser** complains about `bias` is not an attribute when it parses the forward function that uses `self.bias`, no matter what the __getstate__ and __setstate__ do (as long as they exist).

File: torch/utils/mkldnn.py [hunk 7]
```diff
@@ -1,43 +1,160 @@
...
+        else:
+            self.has_bias = torch.jit.Attribute(False, bool)
+            # TODO: Remove this once ScriptModule supports registering None buffer
+            self.register_buffer(
+                'bias',
+                torch.zeros([dense_module.weight.size(0)], dtype=torch.float).to_mkldnn())
+
+    @torch.jit.script_method
+    def __getstate__(self):
+        return (self.weight.to_dense(), self.bias.to_dense(), self.has_bias)
```
Comment (dzhulgakov) [inline torch/utils/mkldnn.py]: I wish there was a workaround that require adding additional field. Let's land it for now but please create a lockdown issue with example code you want to have  working, ok?

**Review by dzhulgakov — APPROVED** [review]

File: torch/utils/mkldnn.py [hunk 8]
```diff
@@ -1,43 +1,149 @@
...
 
 
-def to_mkldnn(module):
-    def t_fn(t):
-        if t.is_floating_point():
-            return t.to_mkldnn()
+class MkldnnLinear(torch.jit.ScriptModule):
+    def __init__(self, dense_module):
+        super(MkldnnLinear, self).__init__()
+        self.register_buffer('weight', dense_module.weight.to_mkldnn())
```
Comment (XiaobingSuper) [inline torch/utils/mkldnn.py]: @bddppq , I doubt that why we regist weight to a buffer not a parameter, it is not suitable to training a mkldnn module if the weight is regist as a buffer. Can you tell me when we will use the jit save path?

Comment (XiaobingSuper) [reply torch/utils/mkldnn.py]: @bddppq , I have tried regist weight to a parameter to run backward, I found backward operation can be run, but the jit save and load have some problem, can you give me some advice? Thanks!
